azkaban-exec-server/src/test/java/azkaban/container/FlowContainerTest.java

```diff
@@ -99,23 +100,47 @@ public void setup() throws Exception {
     this.projectLoader = mock(ProjectLoader.class);
   }
 
-  /**
-   * FIXME: This test is incomplete for now as there is an expected merge conflict which will
-   * change interface with submitFlow. Once new code is merged in FlowContainer,
-   * this test will need to be completed by invoking a flow.
-   * @throws Exception
-   */
-  @Test
-  public void testExecSimple() throws Exception {
+  @After
+  public void destroy() {
+    if (this.flowContainer == null) {
+      return;
+    }
+    try {
+      this.flowContainer.closeMBeans();
+      this.flowContainer = null;
+    } catch (final Exception e) {
+      e.printStackTrace();
+    }
+  }
+
+  private void startFlowContainer() throws IOException {
     final ExecutableFlow execFlow = createTestExecutableFlowFromYaml("basicflowyamltest", "basic_flow");
     execFlow.setExecutionId(1);
     final ProjectFileHandler handler = new ProjectFileHandler(1, 1, 1, "testUser", "zip", "test.zip",
         1, null, null, null, "111.111.111.111");
     when(this.projectLoader.fetchProjectMetaData(anyInt(), anyInt())).thenReturn(handler);
 
-    final FlowContainer flowContainer = SERVICE_PROVIDER.getInstance(FlowContainer.class);
-    flowContainer.start(props);
-    FlowContainer.launchCtrlMsgListener(flowContainer);
+    this.flowContainer = SERVICE_PROVIDER.getInstance(FlowContainer.class);
+    this.flowContainer.start(props);
+  }
+
+  /**
+   * Don't like the name? Me neither - DJ
+   * Test if Callback Manager is not created if not enabled.
+   * This test must be the first as Callback Manager is a static object which once created
```

Comment:
Looks like `JobCallbackManager` is creating a singleton the old-fashioned way (with a static `instance` variable), with some fairly obvious drawbacks, especially for testability.
One approach for resolving this would be to make it a Guice based `@Singleton`, that you can freely instantiate for testing.
Another simple (but less preferred) alternative, make this consturctor `private JobCallbackManager(final Props props)` either package-private or `public`, annotate it with `@VisibleForTesting`  and use it directly to instantiate more objects of type `JobCallbackManager`.
I think both of the above approaches would be simpler (and better) than having to jump through some of these hoops for ensuring an ordering of test cases.

Reply:
If the callback is not enabled then JobCallbackManager should not be instantiated, I think that is why it is not Guice based.
The jobCallbackManager is created from within the Azkaban code and not directly from the test code. Unless we change the Azkaban code to make it happen we cant utilize the 2nd approach.

That is too much code change to make the tests work.
The current approach has only condition which is much easier to meet and avoids touching the existing code.